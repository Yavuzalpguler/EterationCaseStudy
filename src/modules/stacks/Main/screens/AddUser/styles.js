import React from 'react';
import {StyleSheet} from 'react-native';
import {
  heightPercentage,
  widthPercentage,
} from '../../../../../constants/StyleVariables';
const styles = StyleSheet.create({
  textInput: {
    width: '90%',
    height: widthPercentage(13),

    alignSelf: 'center',
    marginTop: heightPercentage(1),
    color: 'black',
    fontSize: widthPercentage(3.5),
    paddingLeft: widthPercentage(3),
    backgroundColor: 'white',
    borderRadius: widthPercentage(2),
    borderWidth: 2,
    borderColor: 'rgba(229, 229, 229, 0.9)',
  },
  addCharButton: {
    width: widthPercentage(90),
    height: widthPercentage(13),
    alignSelf: 'center',
    marginTop: heightPercentage(5),
    backgroundColor: 'rgb(46,134,222)',
    borderRadius: widthPercentage(2),
    justifyContent: 'center',
    alignItems: 'center',
  },
});
export default styles;
