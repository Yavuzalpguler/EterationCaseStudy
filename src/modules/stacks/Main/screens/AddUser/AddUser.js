import {
  View,
  Text,
  TextInput,
  TouchableOpacity,
  SafeAreaView,
  Alert,
} from 'react-native';
import React, {useEffect, useState, useRef} from 'react';
import {
  heightPercentage,
  widthPercentage,
} from '../../../../../constants/StyleVariables';
import styles from './styles';
import {useDispatch, useSelector} from 'react-redux';
import {updateMainCustomState} from '../../../+store/Main/Main.action';

const AddUser = props => {
  const [nameSurname, setNameSurname] = useState('');
  const [jobTitle, setJobTitle] = useState('');
  const [about, setAbout] = useState('');
  const [imageLink, setImageLink] = useState('');
  const [focusedIndex, setFocusedIndex] = useState(0);
  const dispatch = useDispatch();
  const userInfo = useSelector(state => state.main.userInfo);
  const textRef1 = useRef();
  const textRef2 = useRef();
  const textRef3 = useRef();
  const textRef4 = useRef();

  useEffect(() => {
    if (focusedIndex === 0) {
      textRef1.current.focus();
    }
    if (focusedIndex === 1) {
      textRef2.current.focus();
    }
    if (focusedIndex === 2) {
      textRef3.current.focus();
    }
    if (focusedIndex === 3) {
      textRef4.current.focus();
    }
  }, [focusedIndex]);

  const renderInput = (placeholder, value, setValue, index, referance) => {
    return (
      <View style={{marginTop: widthPercentage(5)}}>
        <Text style={{marginLeft: widthPercentage(5)}}>{placeholder}</Text>
        <TextInput
          selectionColor={'black'}
          style={[
            styles.textInput,
            {
              borderColor:
                focusedIndex === index
                  ? 'rgb(46,134,222)'
                  : 'rgba(229, 229, 229, 0.9)',
            },
          ]}
          ref={referance}
          keyboardType="default"
          autoCapitalize="none"
          value={value}
          onChangeText={text => setValue(text)}
          onFocus={() => {
            setFocusedIndex(index);
          }}
          onSubmitEditing={() => {
            setFocusedIndex(index + 1);
          }}
        />
      </View>
    );
  };

  const handleAddUser = () => {
    if (nameSurname === '' || jobTitle === '' || about === '') {
      Alert.alert('Please fill all fields');
      return;
    }
    dispatch(
      updateMainCustomState('userInfo', {
        data: [
          {
            id: Math.random(),
            name: nameSurname,
            job: jobTitle,
            description: about,
            image: imageLink,
          },
          ...userInfo.data,
        ],
      }),
    );
    Alert.alert;
    alert('Successfully added');
    props.navigation.navigate('MainScreen');
  };

  return (
    <SafeAreaView style={{backgroundColor: 'rgba(242, 242, 242, 1)', flex: 1}}>
      {renderInput('Name Surname:', nameSurname, setNameSurname, 0, textRef1)}
      {renderInput('Job Title:', jobTitle, setJobTitle, 1, textRef2)}
      {renderInput('About Him/Her:', about, setAbout, 2, textRef3)}
      {renderInput('Image Link:', imageLink, setImageLink, 3, textRef4)}
      <TouchableOpacity
        style={styles.addCharButton}
        onPress={() => handleAddUser()}>
        <Text
          style={{
            fontSize: widthPercentage(4),
            color: 'white',
          }}>
          Add Character
        </Text>
      </TouchableOpacity>
    </SafeAreaView>
  );
};

export default AddUser;
