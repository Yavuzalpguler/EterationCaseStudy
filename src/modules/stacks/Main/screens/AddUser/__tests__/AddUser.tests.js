import React from 'react';
import {render, fireEvent, waitFor} from '@testing-library/react-native';
import AddUser from '../AddUser';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';
const mockStore = configureStore([]);
const store = mockStore({
  main: {
    userInfo: {
      data: [
        {
          id: 1,
          name: 'Homer Simpson',
          age: 39,
          image: 'https://i.imgur.com/1TQOw9A.png',
        },
        {
          id: 2,
          name: 'Marge Simpson',
          age: 36,
          image: 'https://i.imgur.com/1TQOw9A.png',
        },
      ],
    },
  },
});
const navigation = {
  navigate: jest.fn(),
};
describe('AddUser', () => {
  //clear mock store actions before each test
  beforeEach(() => {
    store.clearActions();
  });
  //clear inputs after each test
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', () => {
    const {toJSON} = render(
      <Provider store={store}>
        <AddUser />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });
  it('should render correctly with empty data', () => {
    const store = mockStore({
      main: {
        userInfo: {
          data: [],
        },
      },
    });
    const {toJSON} = render(
      <Provider store={store}>
        <AddUser />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should handle add user when all fields are filled', async () => {
    const {getByTestId} = render(
      <Provider store={store}>
        <AddUser navigation={navigation} />
      </Provider>,
    );
    const nameInput = getByTestId('nameInput');
    const jobInput = getByTestId('jobInput');
    const aboutInput = getByTestId('aboutInput');
    const imageInput = getByTestId('imageInput');
    const addButton = getByTestId('addButton');
    fireEvent.changeText(nameInput, 'Homer Simpson');
    fireEvent.changeText(jobInput, 39);
    fireEvent.changeText(aboutInput, 'Dad to all');
    fireEvent.changeText(imageInput, 'https://i.imgur.com/1TQOw9A.png');
    fireEvent.press(addButton);
    await waitFor(() => {
      expect(navigation.navigate).toHaveBeenCalledWith('MainScreen');
      expect(nameInput.props.value).toEqual('Homer Simpson');
      expect(jobInput.props.value).toEqual(39);
      expect(aboutInput.props.value).toEqual('Dad to all');
      expect(imageInput.props.value).toEqual('https://i.imgur.com/1TQOw9A.png');
    });
  });

  it('should not handle add user when all fields are not filled', async () => {
    const {getByTestId} = render(
      <Provider store={store}>
        <AddUser navigation={navigation} />
      </Provider>,
    );
    const nameInput = getByTestId('nameInput');
    const jobInput = getByTestId('jobInput');
    const aboutInput = getByTestId('aboutInput');
    const imageInput = getByTestId('imageInput');
    const addButton = getByTestId('addButton');
    fireEvent.changeText(nameInput, '');
    fireEvent.changeText(jobInput, '');
    fireEvent.changeText(aboutInput, '');
    fireEvent.changeText(imageInput, '');

    fireEvent.press(addButton);
    await waitFor(() => {
      expect(navigation.navigate).not.toHaveBeenCalledWith('MainScreen');
    });
  });

  it('should update store when add user is handled', async () => {
    const {getByTestId} = render(
      <Provider store={store}>
        <AddUser navigation={navigation} />
      </Provider>,
    );
    const nameInput = getByTestId('nameInput');
    const jobInput = getByTestId('jobInput');
    const aboutInput = getByTestId('aboutInput');
    const imageInput = getByTestId('imageInput');
    const addButton = getByTestId('addButton');
    fireEvent.changeText(nameInput, 'Homer Simpson');
    fireEvent.changeText(jobInput, 39);
    fireEvent.changeText(aboutInput, 'Dad to all');
    fireEvent.changeText(imageInput, 'https://i.imgur.com/1TQOw9A.png');
    fireEvent.press(addButton);
    await waitFor(() => {
      expect(store.getActions()).toEqual([
        {
          key: 'userInfo',
          type: '@UPDATE_MAIN_CUSTOM_STATE',
          value: {
            data: [
              {
                description: 'Dad to all',
                id: 3,
                image: 'https://i.imgur.com/1TQOw9A.png',
                job: 39,
                name: 'Homer Simpson',
              },
              {
                id: 1,
                name: 'Homer Simpson',
                age: 39,
                image: 'https://i.imgur.com/1TQOw9A.png',
              },
              {
                id: 2,
                name: 'Marge Simpson',
                age: 36,
                image: 'https://i.imgur.com/1TQOw9A.png',
              },
            ],
          },
        },
      ]);
    });
  });
});
