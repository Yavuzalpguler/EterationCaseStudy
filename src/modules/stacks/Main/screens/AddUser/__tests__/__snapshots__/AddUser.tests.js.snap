// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddUser should render correctly 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "backgroundColor": "rgba(242, 242, 242, 1)",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "marginTop": 37.5,
      }
    }
  >
    <Text
      style={
        Object {
          "marginLeft": 37.5,
        }
      }
    >
      Name Surname:
    </Text>
    <TextInput
      autoCapitalize="none"
      keyboardType="default"
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      selectionColor="black"
      style={
        Array [
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderColor": "rgba(229, 229, 229, 0.9)",
            "borderRadius": 15,
            "borderWidth": 2,
            "color": "black",
            "fontSize": 26.250000000000004,
            "height": 97.5,
            "marginTop": 13.34,
            "paddingLeft": 22.5,
            "width": "90%",
          },
          Object {
            "borderColor": "rgb(46,134,222)",
          },
        ]
      }
      testID="nameInput"
      value=""
    />
  </View>
  <View
    style={
      Object {
        "marginTop": 37.5,
      }
    }
  >
    <Text
      style={
        Object {
          "marginLeft": 37.5,
        }
      }
    >
      Job Title:
    </Text>
    <TextInput
      autoCapitalize="none"
      keyboardType="default"
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      selectionColor="black"
      style={
        Array [
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderColor": "rgba(229, 229, 229, 0.9)",
            "borderRadius": 15,
            "borderWidth": 2,
            "color": "black",
            "fontSize": 26.250000000000004,
            "height": 97.5,
            "marginTop": 13.34,
            "paddingLeft": 22.5,
            "width": "90%",
          },
          Object {
            "borderColor": "rgba(229, 229, 229, 0.9)",
          },
        ]
      }
      testID="jobInput"
      value=""
    />
  </View>
  <View
    style={
      Object {
        "marginTop": 37.5,
      }
    }
  >
    <Text
      style={
        Object {
          "marginLeft": 37.5,
        }
      }
    >
      About Him/Her:
    </Text>
    <TextInput
      autoCapitalize="none"
      keyboardType="default"
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      selectionColor="black"
      style={
        Array [
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderColor": "rgba(229, 229, 229, 0.9)",
            "borderRadius": 15,
            "borderWidth": 2,
            "color": "black",
            "fontSize": 26.250000000000004,
            "height": 97.5,
            "marginTop": 13.34,
            "paddingLeft": 22.5,
            "width": "90%",
          },
          Object {
            "borderColor": "rgba(229, 229, 229, 0.9)",
          },
        ]
      }
      testID="aboutInput"
      value=""
    />
  </View>
  <View
    style={
      Object {
        "marginTop": 37.5,
      }
    }
  >
    <Text
      style={
        Object {
          "marginLeft": 37.5,
        }
      }
    >
      Image Link:
    </Text>
    <TextInput
      autoCapitalize="none"
      keyboardType="default"
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      selectionColor="black"
      style={
        Array [
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderColor": "rgba(229, 229, 229, 0.9)",
            "borderRadius": 15,
            "borderWidth": 2,
            "color": "black",
            "fontSize": 26.250000000000004,
            "height": 97.5,
            "marginTop": 13.34,
            "paddingLeft": 22.5,
            "width": "90%",
          },
          Object {
            "borderColor": "rgba(229, 229, 229, 0.9)",
          },
        ]
      }
      testID="imageInput"
      value=""
    />
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "rgb(46,134,222)",
        "borderRadius": 15,
        "height": 97.5,
        "justifyContent": "center",
        "marginTop": 66.7,
        "opacity": 1,
        "width": 675,
      }
    }
    testID="addButton"
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 30,
        }
      }
    >
      Add Character
    </Text>
  </View>
</RCTSafeAreaView>
`;

exports[`AddUser should render correctly with empty data 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "backgroundColor": "rgba(242, 242, 242, 1)",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "marginTop": 37.5,
      }
    }
  >
    <Text
      style={
        Object {
          "marginLeft": 37.5,
        }
      }
    >
      Name Surname:
    </Text>
    <TextInput
      autoCapitalize="none"
      keyboardType="default"
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      selectionColor="black"
      style={
        Array [
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderColor": "rgba(229, 229, 229, 0.9)",
            "borderRadius": 15,
            "borderWidth": 2,
            "color": "black",
            "fontSize": 26.250000000000004,
            "height": 97.5,
            "marginTop": 13.34,
            "paddingLeft": 22.5,
            "width": "90%",
          },
          Object {
            "borderColor": "rgb(46,134,222)",
          },
        ]
      }
      testID="nameInput"
      value=""
    />
  </View>
  <View
    style={
      Object {
        "marginTop": 37.5,
      }
    }
  >
    <Text
      style={
        Object {
          "marginLeft": 37.5,
        }
      }
    >
      Job Title:
    </Text>
    <TextInput
      autoCapitalize="none"
      keyboardType="default"
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      selectionColor="black"
      style={
        Array [
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderColor": "rgba(229, 229, 229, 0.9)",
            "borderRadius": 15,
            "borderWidth": 2,
            "color": "black",
            "fontSize": 26.250000000000004,
            "height": 97.5,
            "marginTop": 13.34,
            "paddingLeft": 22.5,
            "width": "90%",
          },
          Object {
            "borderColor": "rgba(229, 229, 229, 0.9)",
          },
        ]
      }
      testID="jobInput"
      value=""
    />
  </View>
  <View
    style={
      Object {
        "marginTop": 37.5,
      }
    }
  >
    <Text
      style={
        Object {
          "marginLeft": 37.5,
        }
      }
    >
      About Him/Her:
    </Text>
    <TextInput
      autoCapitalize="none"
      keyboardType="default"
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      selectionColor="black"
      style={
        Array [
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderColor": "rgba(229, 229, 229, 0.9)",
            "borderRadius": 15,
            "borderWidth": 2,
            "color": "black",
            "fontSize": 26.250000000000004,
            "height": 97.5,
            "marginTop": 13.34,
            "paddingLeft": 22.5,
            "width": "90%",
          },
          Object {
            "borderColor": "rgba(229, 229, 229, 0.9)",
          },
        ]
      }
      testID="aboutInput"
      value=""
    />
  </View>
  <View
    style={
      Object {
        "marginTop": 37.5,
      }
    }
  >
    <Text
      style={
        Object {
          "marginLeft": 37.5,
        }
      }
    >
      Image Link:
    </Text>
    <TextInput
      autoCapitalize="none"
      keyboardType="default"
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      selectionColor="black"
      style={
        Array [
          Object {
            "alignSelf": "center",
            "backgroundColor": "white",
            "borderColor": "rgba(229, 229, 229, 0.9)",
            "borderRadius": 15,
            "borderWidth": 2,
            "color": "black",
            "fontSize": 26.250000000000004,
            "height": 97.5,
            "marginTop": 13.34,
            "paddingLeft": 22.5,
            "width": "90%",
          },
          Object {
            "borderColor": "rgba(229, 229, 229, 0.9)",
          },
        ]
      }
      testID="imageInput"
      value=""
    />
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "rgb(46,134,222)",
        "borderRadius": 15,
        "height": 97.5,
        "justifyContent": "center",
        "marginTop": 66.7,
        "opacity": 1,
        "width": 675,
      }
    }
    testID="addButton"
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 30,
        }
      }
    >
      Add Character
    </Text>
  </View>
</RCTSafeAreaView>
`;
